### ISSUE TITLE
Incorrect behavior of AltGr+é on French keyboard

### ISSUE THREAD

**Issue by sba923** [PowerShell/PSReadLine#1226]
This is a new issue with 2.0.0 beta 6, whereas all versions since beta 4 seemed to have fixed the earlier AltGr issues (whatever the host: Windows PowerShell 5.1, PowerShell Core 6.2.3, PowerShell 7.0.0 previews, outside or inside VScode).

Unfortunately, this seems to be occurring randomly; I don't have a 100% reliable way to repro it--for instance it's currently present in one of my PowerShell 7.0.0 preview 6 instances, a new one (exact same profile, etc.) isn't affected.

Environment data
----------------

PS version: 7.0.0-preview.6
PSReadline version: 2.0.0-beta6
os: 10.0.17134.1121 (WinBuild.160101.0800)
PS file version: 7.0.0.0
HostName: ConsoleHost
BufferWidth: 120
BufferHeight: 9001

Steps to reproduce or exception report
--------------------------------------
When I hit `AltGr+é` on my French keyboard I should get... nothing, until I then hit `Space` which then yields `~`

What occurs when the issue shows up is that `AltGr+é` yields `é`, and when I subsequently hit `Space` I get `~`. Bottom line is that the input is `é~` instead of the expected `~`.

**Comment by daxian-dbw** [comment]
I just reviewed the changes included in 2.0.0-beta6 and none of them would affect the dead key handling. Are you sure this only happens with the 2.0.0-beta6 PSReadLine?

**Comment by sba923** [comment]
No, I'm not positive. Do you want me to downgrade that particular system to, say, beta5 and check what I get from there?

**Comment by daxian-dbw** [comment]
Yes, that would be very helpful. The dead key fix was made in 2.0.0-beta5 (https://github.com/PowerShell/PSReadLine/commit/f2f553156063a8635b9d4882a878a828365cdd99).

**Comment by sba923** [comment]
Where can I get the bits for 2.0.0-beta5 or even 2.0.0-beta4?

**Comment by daxian-dbw** [comment]
@sba923 You can get the previous versions from PowerShell Gallery, by specifying the required version via `-RequiredVersion`.

**Comment by sba923** [comment]
How?

![image](https://user-images.githubusercontent.com/12860484/70424262-62d46480-1a6f-11ea-98ba-d32ff552967c.png)

**Comment by daxian-dbw** [comment]
From within a 7.0.0-preview.6 session, the 2.0.0-beta5 can be installed by
```
Install-Module PSReadLine -RequiredVersion 2.0.0-beta5 -AllowPrerelease
```
The module will be installed at `C:\Users\<user>\Documents\PowerShell\Modules`. This module path `C:\Users\<user>\Documents\PowerShell\` is placed at the first in `$env:PSModulePath`, so it would be picked up by pwsh.

**Comment by sba923** [comment]
This is driving me nuts:

![image](https://user-images.githubusercontent.com/12860484/70470319-0737c480-1acb-11ea-82ad-d05d0c92e5a3.png)

**Comment by msftrncs** [comment]
You can attempt `-Force` but I find I still get this error.

I am pretty sure it happens because the system install of PSReadLine is in use, even when you think it isn't.  You might have to just delete the system install with a file explorer after closing all instances of PowerShell.

**Comment by daxian-dbw** [comment]
This happens because you already have a pre-release version of PSReadLine module installed at `C:\Users\<user>\Documents\PowerShell\Modules`. `PowerShellGet` and `PowerShell` doesn't handle semantic version of modules well enough. For different stable versions, the modules will be installed side by side, but for pre-release versions such as 2.0.0-betaX, they will always be installed in the folder `2.0.0`. In this case, there is already a `2.0.0-beta6` installed at `C:\Users\<user>\Documents\PowerShell\Modules\PSReadLine\2.0.0`, so the installation of `2.0.0-beta5` would attempt to delete files in that folder. However, the `PSReadLine2.dll` has already been loaded in the same process and thus cannot be removed, hence the error.

Here is the workaround:
1. close all pwsh sessions
2. from cmd, start pwsh by `pwsh -noprofile -noninteractive`. PSReadLine will not be loaded by having `-noninteractive`
3. then run `Install-Module PSReadLine -RequiredVersion 2.0.0-beta5 -AllowPrerelease`

**Comment by sba923** [comment]
Yes @SteveL-MSFT already told me about that semantic versioning limitation in another thread.

Without -Force I get:

![image](https://user-images.githubusercontent.com/12860484/70530067-570d9e80-1b52-11ea-8776-d9c4f14f280f.png)

-Force gets me past this and now I have:

![image](https://user-images.githubusercontent.com/12860484/70530198-a2c04800-1b52-11ea-9ea2-eb8a88f42da4.png)

Will get rid of all copies of 2.0.0-beta6 to be sure. I'll keep you posted.

**Comment by sba923** [comment]
Now that machine should have nothing but 2.0.0-beta5:

![image](https://user-images.githubusercontent.com/12860484/70536782-ab6b4b00-1b5f-11ea-9374-7e6ee9d65ffb.png)

I'll keep looking for the `AltGr+é` issue.

(I'm assuming renamed folders now called `PSReadLine-2.0.0beta6` will never be picked up, is that a fair assumption?)

**Comment by daxian-dbw** [comment]
> I'm assuming renamed folders now called PSReadLine-2.0.0beta6 will never be picked up, is that a fair assumption?

Yes, it won't be picked up, as PowerShell's module discovery doesn't handle the semantics version yet :(

Thanks for spending all the efforts, I really appreciate it!

**Comment by sba923** [comment]
This worked, but only for PS6/7. For Windows PowerShell whose `Install-Module` doesn't support `-AllowPrerelease`, I've stuck with the dirty "copy PSRL over from some other location" trick ;-).

I'll stick with `beta5` for a while on that machine, refrain from going to  `rc1` and keep monitoring for occurrences of the `AltGr+é` issue.

**Comment by daxian-dbw** [comment]
For Windows PowerShell, you will need to upgrade PowerShellGet first. Please see the instructions here: https://github.com/powershell/psreadline#install-from-powershellgallery-preferred

**Comment by sba923** [comment]
Haven't experienced the issue since I switched back to beta5. Want me to hop to rc1 and monitor?

**Comment by daxian-dbw** [comment]
@sba923 Good to know that! Yes, please switch to rc1 and see if the issue appears again. You can install the rc1 this way:

1. close all pwsh sessions
2. from cmd, start pwsh by `pwsh -noprofile -noninteractive`. PSReadLine will not be loaded by having `-noninteractive`
3. then run `Install-Module PSReadLine -RequiredVersion 2.0.0-rc1 -AllowPrerelease`

Please let me know if you run into the issue again with rc1. Thanks!

**Comment by msftrncs** [comment]
@daxian-dbw, I think you might have meant for the step 3 line to not include the `-RequiredVersion` parameter/value, in order to update to the latest version.

**Comment by daxian-dbw** [comment]
@msftrncs I copied those 3 lines from an above comment and forgot to update the version. Thanks for pointing out!

**Comment by sba923** [comment]
I get the issue with PSReadLine 2.0.0-rc1 in PowerShell 7.0.0-rc1 running within VSCode with version 2019.12.0 of the PowerShell Preview extension.

I don't get the issue (so far) with PSReadLine 2.0.0-rc1 in PowerShell 7.0.0-rc1 running in its own console (the "standard" way).

**Comment by daxian-dbw** [comment]
@sba923 Does it happen consistently in VSCode? It would be really helpful if there is a repro.

**Comment by sba923** [comment]
It seems it does on that particular machine (my work laptop)...

**Comment by sba923** [comment]
Here's what I get in VScode (PS7.0.0-rc.1, PSReadLine 2.0.0-rc1) if I press and hold `AltGr` and then repeatedly press `é`:

![image](https://user-images.githubusercontent.com/12860484/71807758-50e4e200-306c-11ea-9e33-65654149f2d7.png)

**Comment by daxian-dbw** [comment]
Thanks @sba923, I will try reproducing it myself.

**Comment by daxian-dbw** [comment]
@sba923 I'm able to reproduce the issue in VS Code after unloading `PSReadLine` as well as using `CMD` as the shell (see the screenshot below).
I tried in Windows Terminal too and it works fine there. So it looks to me an issue specific to VS Code.

![image](https://user-images.githubusercontent.com/127450/71838340-732b2f80-306d-11ea-88ea-21b17bb53ee2.png)

I opened an issue in VSCode repo: https://github.com/microsoft/vscode/issues/88156, and will close this one as external.
